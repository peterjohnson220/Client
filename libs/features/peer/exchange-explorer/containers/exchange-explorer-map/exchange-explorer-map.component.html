<mgl-map
  [style]="mapStyle"
  [attributionControl]="false"
  [fitBounds]="peerInitialMapBounds$ | async"
  [fitBoundsOptions]="{padding: 50, duration: 500 }"
  [center]="peerMapCentroid$ | async"
  [zoom]="peerMapInitialZoomLevel$ | async"
  [maxZoom]="peerMapMaxZoom$ | async"
  [cursorStyle]="cursorStyle"
  [renderWorldCopies]="false"
  [maxBounds]="[[-180, -85], [180, 85]]"
  (moveEnd)="handleMoveEndEvent($event)"
  (load)="handleLoadEvent($event)"
  (zoomEnd)="handleZoomEnd($event)"
  (resize)="handleResizeEvent()"
  [ngClass]="{'map-no-data': (peerMapShowNoData$ | async)}"
>

  <mgl-control mglGeocoder></mgl-control>
  <mgl-control mglNavigation></mgl-control>

  <mgl-control>

    <div class="mapboxgl-ctrl-group">
      <button
        class="btn btn-secondary"
        (click)="toggleSatelliteStyle()"
        [title]="satelliteStyleEnabledText"
      >
        <span *ngIf="!satelliteStyleEnabled" ><fa-icon icon="globe"></fa-icon></span>
        <span *ngIf="satelliteStyleEnabled" ><fa-icon [icon]="['far', 'map']"></fa-icon></span>
      </button>
    </div>

  </mgl-control>

  <!--Data Source-->
  <mgl-geojson-source
    [id]="'incumbents'"
    [data]="peerMapCollection$ | async">
  </mgl-geojson-source>

  <!--Selected Point Info-->
  <mgl-popup
    *ngIf="selectedPoint"
    [lngLat]="selectedPoint.geometry.coordinates"
    (close)="selectedPoint = null"
    [closeButton]="false"
  >
    <div class="point-information">
      <div class="point-information-header">
        <strong>{{selectedPoint.properties.CompanyCount}}</strong> Orgs with
        <strong>{{selectedPoint.properties.SalaryCount}}</strong> Incs in this area
      </div>

      <div class="point-information-body">
        <ng-container *ngFor="let company of pointCompaniesPreview">

          <div class="d-flex justify-content-between">
            <div>{{company.Company}}</div>
            <div class="incumbent-count">{{company.Count}}</div>
          </div>

        </ng-container>
      </div>

    </div>
  </mgl-popup>

  <!--Multiple Company OR Multiple LatLon Circle Layer-->
  <mgl-layer
    id="multiple-company-latlon-layer"
    type="circle"
    source="incumbents"
    (click)="handleLayerClusteredClickEvent($event)"
    (mouseMove)="handleLayerHoverEvent($event)"
    (mouseEnter)="cursorStyle = 'pointer'"
    (mouseLeave)="cursorStyle = ''; selectedPoint = null"
    [paint]="{
        'circle-color': '#3f8845',
        'circle-radius': [
                'interpolate', ['linear'], ['zoom'],
                4,
                12,
                9,
                8
              ],
        'circle-stroke-width': 1,
        'circle-stroke-color': '#ffffff'
      }"
    [filter]="['any', ['>', 'CompanyCount', 1], ['>', 'LatLonCount', 1]]"
  >
  </mgl-layer>

  <!--Single Company/LatLon Layer-->
  <mgl-layer
    id="single-company-latlon-layer"
    type="circle"
    source="incumbents"
    (click)="handleLayerClusteredClickEvent($event)"
    (mouseMove)="handleLayerHoverEvent($event)"
    (mouseEnter)="cursorStyle = 'pointer'"
    (mouseLeave)="cursorStyle = ''; selectedPoint = null"
    [paint]="{
        'circle-color': '#3f8845',
        'circle-radius': 7,
        'circle-stroke-width': 1,
        'circle-stroke-color': '#ffffff'
      }"
    [filter]="['all', ['==', 'CompanyCount', 1], ['==', 'LatLonCount', 1]]"
  >
  </mgl-layer>

  <!--Multiple Company OR Multiple LatLon Count Layer-->
  <mgl-layer
    id="cluster-count-layer"
    type="symbol"
    source="incumbents"
    [layout]="{
        'text-field': '{CompanyCount}',
        'text-font': ['DIN Offc Pro Medium', 'Arial Unicode MS Bold'],
        'text-size': 10
      }"
    [paint]="{
      'text-color': '#ffffff'
    }"
    [filter]="['any', ['>', 'CompanyCount', 1], ['>', 'LatLonCount', 1]]"
  >
  </mgl-layer>

  <!--No Data Popup-->
  <mgl-popup
    *ngIf="(peerMapShowNoData$ | async) && !(peerMapAutoZooming$ | async) && (peerMapLoaded$ | async) && !(dataLoading$ | async)"
    [lngLat]="center"
    [closeOnClick]="false"
    [closeButton]="false"
  >
    <div class="map-no-data-message">
      <fa-icon [icon]="['far','exclamation-triangle']" class="text-danger"></fa-icon>
      No results. Refine the filter criteria to show results.
    </div>
  </mgl-popup>
</mgl-map>
