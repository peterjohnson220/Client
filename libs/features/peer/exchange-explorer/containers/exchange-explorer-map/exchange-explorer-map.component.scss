@import 'colors';

:host{
  height: 100%;
}

mgl-map {
  height: 100%;
  width: 100%;
}

:host ::ng-deep .map-no-data .mapboxgl-popup-anchor-bottom .mapboxgl-popup-tip {
  display: none !important;
}

.map-no-data-message {
  font-size: 1.5rem;
  padding: 1rem;
}

:host ::ng-deep .mapboxgl-popup {
  max-width: none !important;

  .mapboxgl-popup-content {
    padding: 0;
  }
}

.point-information {
  font-size: 1rem;
  width: 260px;
}

.point-information-header {
  border-bottom: 1px solid $border-grey;
  background-color: #EDEDED;
  padding: .5rem 1rem;
}

.point-information-body {
  padding: 1rem;
  background-color: #FFFFFF;
}

.incumbent-count {
  color: #9a9a9a;
}

:host ::ng-deep .mapboxgl-ctrl-geocoder--input {
  font-size: 1rem;
}
