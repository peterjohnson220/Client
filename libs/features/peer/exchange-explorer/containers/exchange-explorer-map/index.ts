export * from './exchange-explorer-map.component';
