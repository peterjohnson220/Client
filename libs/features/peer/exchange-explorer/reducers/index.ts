import { createSelector, createFeatureSelector } from '@ngrx/store';

// Import root app reducer
import * as fromRoot from 'libs/state/state';

// Import feature reducers
import * as fromExchangeScopeReducer from './exchange-scope.reducer';
import * as fromExchangeFilterContextReducer from './exchange-filter-context.reducer';
import * as fromMapReducer from './map.reducer';
import * as fromExchangeExplorerContextInfoReducer from './exchange-explorer-context-info.reducer';
import * as fromExchangeDataCutReducer from './exchange-data-cut.reducer';

// Feature area state
export interface FeaturePeerExchangeExplorerState {
  map: fromMapReducer.State;
  exchangeScope: fromExchangeScopeReducer.State;
  exchangeFilterContext: fromExchangeFilterContextReducer.State;
  exchangeExplorerContextInfo: fromExchangeExplorerContextInfoReducer.State;
  exchangeDataCut: fromExchangeDataCutReducer.State;
}

// Extend root state with feature area state
export interface State extends fromRoot.State {
  feature_peer_exchangeExplorer: FeaturePeerExchangeExplorerState;
}

// Feature area reducers
export const reducers = {
  map: fromMapReducer.reducer,
  exchangeScope: fromExchangeScopeReducer.reducer,
  exchangeFilterContext: fromExchangeFilterContextReducer.reducer,
  exchangeExplorerContextInfo: fromExchangeExplorerContextInfoReducer.reducer,
  exchangeDataCut: fromExchangeDataCutReducer.reducer
};

// Select Feature Area
export const selectFeaturePeerExchangeExplorerState = createFeatureSelector<FeaturePeerExchangeExplorerState>(
  'feature_peer_exchangeExplorer'
);

// Feature Selectors
export const selectMapState = createSelector(
  selectFeaturePeerExchangeExplorerState,
  (state: FeaturePeerExchangeExplorerState) => !!state ? state.map : {}
);

export const selectExchangeScopeState = createSelector(
  selectFeaturePeerExchangeExplorerState,
  (state: FeaturePeerExchangeExplorerState) => !!state ? state.exchangeScope : <fromExchangeScopeReducer.State>{}
);

export const selectFilterContextState = createSelector(
  selectFeaturePeerExchangeExplorerState,
  (state: FeaturePeerExchangeExplorerState) => !!state ? state.exchangeFilterContext : {}
);

export const selectExchangeExplorerContextInfo = createSelector(
  selectFeaturePeerExchangeExplorerState,
  (state: FeaturePeerExchangeExplorerState) => !!state ? state.exchangeExplorerContextInfo : {}
);

export const selectDataCutState = createSelector(
  selectFeaturePeerExchangeExplorerState,
  (state: FeaturePeerExchangeExplorerState) => !!state ? state.exchangeDataCut : {}
);

// Map Data Selectors
export const getPeerMapLoading = createSelector(selectMapState, fromMapReducer.getLoading);
export const getPeerMapLoadingError = createSelector(selectMapState, fromMapReducer.getLoadingError);
export const getPeerMapSummary = createSelector(selectMapState, fromMapReducer.getMapSummary);
export const getPeerMapCompaniesFromSummary = createSelector(
  getPeerMapSummary,
  (summary) => !!summary ? summary.OverallMapStats.Companies : []
);
export const getPeerMapUntaggedIncumbentCount = createSelector(
  getPeerMapSummary,
  (summary) => !!summary ? summary.OverallMapStats.UntaggedIncumbentCount : 0
);
export const getPeerMapExchangeJobIds = createSelector(
  getPeerMapSummary,
  (summary) => !!summary ? summary.OverallMapStats.ExchangeJobIds : []
);
export const getPeerMapOrgCount = createSelector(
  getPeerMapSummary,
  (summary) => !!summary ? summary.OverallMapStats.CompanyCount : 0
);
export const getPeerMapFilter = createSelector(selectMapState, fromMapReducer.getMapFilter);
export const getPeerMapCollection = createSelector(selectMapState, fromMapReducer.getMapCollection);
export const getPeerInitialMapBounds = createSelector(selectMapState, fromMapReducer.getInitialMapBounds);
export const getPeerMapMaxZoom = createSelector(selectMapState, fromMapReducer.getMaxZoom);
export const getPeerMapInitialZoomLevel = createSelector(selectMapState, fromMapReducer.getInitialZoomLevel);
export const getPeerMapCentroid = createSelector(selectMapState, fromMapReducer.getMapCentroid);
export const canLoadPeerMap = createSelector(selectMapState, fromMapReducer.canLoadMap);
export const peerMapShowNoData = createSelector(selectMapState, fromMapReducer.showNoData);
export const peerMapHasUntaggedIncumbents = createSelector(selectMapState, fromMapReducer.hasUntaggedIncumbents);
export const getPeerMapApplyingScope = createSelector(selectMapState, fromMapReducer.getApplyingScope);
export const getPeerMapAutoZooming = createSelector(selectMapState, fromMapReducer.getAutoZooming);
export const getPeerMapZoom = createSelector(selectMapState, fromMapReducer.getZoom);
export const getPeerMapInitialZoomComplete = createSelector(selectMapState, fromMapReducer.getInitialZoomComplete);
export const getPeerMapLoaded = createSelector(selectMapState, fromMapReducer.getMapLoaded);

// Exchange Scope Selectors
export const getExchangeScopesLoadingByJobs = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getLoadingByJobs);
export const getExchangeScopesLoadingByJobsError = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getLoadingByJobsError);
export const getExchangeScopesLoadingByExchange = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getLoadingByExchange);
export const getExchangeScopesLoadingByExchangeError =
  createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getLoadingByExchangeError);
export const getExchangeScopeDetailsLoading = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getLoadingDetails);
export const getExchangeScopeDetailsLoadingError = createSelector(
  selectExchangeScopeState,
  fromExchangeScopeReducer.getLoadingDetailsError
);
export const getExchangeScopeUpserting = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getUpserting);
export const getExchangeScopeUpsertingError = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getUpsertingError);
export const getDeletingExchangeScope = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getDeletingScope);
export const getDeletingExchangeScopeError = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getDeletingScopeError);
export const getInDeleteExchangeScopeMode = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getInDeleteScopeMode);
export const getExchangeScopeToDelete = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getExchangeScopeToDelete);
export const getIncludeCompanyScopes = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getIncludeCompanyScopes);
export const getIncludeStandardScopes = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getIncludeStandardScopes);
export const getExchangeScopes = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getExchangeScopes);
export const getExchangeScopeNameFilter = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getExchangeScopeNameFilter);
export const getIsForcedByExchange = createSelector(selectExchangeScopeState, fromExchangeScopeReducer.getIsForcedByExchange);

// Exchange Explorer Info Selectors
export const getExchangeExplorerPayMarket = createSelector(
  selectExchangeExplorerContextInfo,
  fromExchangeExplorerContextInfoReducer.getPayMarket
);
export const getSearchFilterMappingDataObj = createSelector(
  selectExchangeExplorerContextInfo,
  fromExchangeExplorerContextInfoReducer.getSearchFilterMappingDataObj
);
export const getExchangeJobFilterOptions = createSelector(
  selectExchangeExplorerContextInfo,
  fromExchangeExplorerContextInfoReducer.getExchangeJobFilterOptions
);
export const getExchangeExplorerPayMarketGeoData = createSelector(
  selectExchangeExplorerContextInfo,
  fromExchangeExplorerContextInfoReducer.getPayMarketGeoData
);
export const getIncludeDisabledFilters = createSelector(
  selectExchangeExplorerContextInfo,
  fromExchangeExplorerContextInfoReducer.getIncludeDisabledFilters
);
export const getIncludeCurrentCompany = createSelector(
  selectExchangeExplorerContextInfo,
  fromExchangeExplorerContextInfoReducer.getIncludeCurrentCompany
);

// Exchange Filter Context Selectors
export const getFilterContextLimitToPayMarket = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getLimitToPayMarket
);
export const getFilterContextExcludeIndirectJobMatches = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getExcludeIndirectJobMatches
);
export const getFilterContextHasSimilarJobLevels = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getHasSimilarJobLevels
);
export const getFilterContextIncludeUntaggedIncumbents = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getIncludeUntaggedIncumbents
);
export const getFilterContextScopeId = createSelector(selectFilterContextState, fromExchangeFilterContextReducer.getScopeId);
export const getFilterContext = createSelector(selectFilterContextState, fromExchangeFilterContextReducer.getFilterContext);
export const getSelectedExchangeJobId = createSelector(
  getFilterContext,
  (filterContext) => {
    if (!filterContext) {
      return 0;
    }
    const exchangeJobId = !!filterContext.LockedExchangeJobId ? filterContext.LockedExchangeJobId : filterContext.ExchangeJobId;
    return !!exchangeJobId ? exchangeJobId : 0;
  }
);
export const getWeightingType = createSelector(selectFilterContextState, fromExchangeFilterContextReducer.getWeightingType);

// Exchange Data Cut Selectors
export const getDataCutLoading = createSelector(selectDataCutState, fromExchangeDataCutReducer.getLoading);

// MISC
export const getAssociatedExchangeJobIds = createSelector(
  getExchangeJobFilterOptions,
  (exchangeJobs) => !!exchangeJobs && !!exchangeJobs.length ? exchangeJobs.map(ej => ej.ExchangeJobId) : []
);
export const getHasAppliedFilterContext = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getHasAppliedContext
);


export const getHasDefaultScope = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getHasDefaultScope
);

export const getExchangeId = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getExchangeId
);

export const mapHasExchangeJobs = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getMapHasExchangeJobs
)

export const getSelectedExchangeScope = createSelector(
  selectFilterContextState,
  fromExchangeFilterContextReducer.getSelectedExchangeScope
);

export const getFilterContextHasDefaultScope = createSelector(
  getHasDefaultScope,
  getFilterContext,
  getExchangeScopes,
  (hasDefaultScope, filterContext, scopes) => {
    if (!hasDefaultScope) {
      return false;
    }

    return !!filterContext && !!scopes && scopes.length && scopes.findIndex(s => s.ExchangeScopeId === filterContext.ScopeId) > -1;
});

export const getFilterContextScopeSelection = createSelector(
  getExchangeScopes,
  getFilterContextScopeId,
  (scopes, selectedId) => {
    if (!scopes || !scopes.length || !selectedId) {
      return null;
    }
    return scopes.find(s => s.ExchangeScopeId === selectedId);
  });
