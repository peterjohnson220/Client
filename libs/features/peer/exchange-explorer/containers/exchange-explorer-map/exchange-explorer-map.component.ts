import { Component, OnDestroy, OnInit } from '@angular/core';

import { select, Store } from '@ngrx/store';
import { Observable, Subscription } from 'rxjs';
import { take } from 'rxjs/operators';
import { FeatureCollection, Point } from 'geojson';
import * as mapboxgl from 'mapbox-gl';
import { LngLatBounds } from 'mapbox-gl';

import { ExchangeMapSummary } from 'libs/models/peer';
import * as fromSearchReducer from 'libs/features/search/reducers';
import * as fromRootReducer from 'libs/state/state';
import { UserContext } from 'libs/models/security';

import * as fromMapActions from '../../actions/map.actions';
import * as fromExchangeExplorerReducer from '../../reducers';

@Component({
  selector: 'pf-exchange-explorer-map',
  templateUrl: './exchange-explorer-map.component.html',
  styleUrls: ['./exchange-explorer-map.component.scss']
})
export class ExchangeExplorerMapComponent implements OnInit, OnDestroy {
  selectedPoint: any = null;
  cursorStyle: string;
  satelliteStyleEnabled = false;
  map: mapboxgl.Map;
  ignoreNextMoveEnd = false;
  peerMapSummary: ExchangeMapSummary;
  peerInitialMapBounds: number[];
  peerInitialMapZoomLevel: number;
  initialMoveEnd = true;
  mbAccessToken: string;

  peerMapCollection$: Observable<FeatureCollection<Point>>;
  peerMapSummary$: Observable<ExchangeMapSummary>;
  dataLoading$: Observable<boolean>;
  peerMapLoadingError$: Observable<boolean>;
  peerMapCentroid$: Observable<number[]>;
  peerMapShowNoData$: Observable<boolean>;
  peerMapHasUntaggedIncumbents$: Observable<boolean>;
  peerMapMaxZoom$: Observable<number>;
  peerInitialMapBounds$: Observable<number[]>;
  peerMapInitialZoomLevel$: Observable<number>;
  peerMapAutoZooming$: Observable<boolean>;
  peerMapInitialZoomComplete$: Observable<boolean>;
  peerMapLoaded$: Observable<boolean>;
  userContext$: Observable<UserContext>;
  peerMapSummarySub: Subscription;
  peerInitialMapBoundsSub: Subscription;
  peerInitialMapZoomLevelSub: Subscription;
  userContextSub: Subscription;

  constructor(private store: Store<fromExchangeExplorerReducer.State>) {
    this.peerMapSummary$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapSummary));
    this.dataLoading$ = this.store.pipe(select(fromSearchReducer.getLoadingResults));
    this.peerMapLoadingError$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapLoadingError));
    this.peerMapCollection$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapCollection));
    this.peerMapShowNoData$ = this.store.pipe(select(fromExchangeExplorerReducer.peerMapShowNoData));
    this.peerMapHasUntaggedIncumbents$ = this.store.pipe(select(fromExchangeExplorerReducer.peerMapHasUntaggedIncumbents));
    this.peerMapMaxZoom$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapMaxZoom));
    this.peerInitialMapBounds$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerInitialMapBounds));
    this.peerMapInitialZoomLevel$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapInitialZoomLevel));
    this.peerMapCentroid$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapCentroid));
    this.peerMapAutoZooming$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapAutoZooming));
    this.peerMapInitialZoomComplete$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapInitialZoomComplete));
    this.peerMapLoaded$ = this.store.pipe(select(fromExchangeExplorerReducer.getPeerMapLoaded));
    this.userContext$ = this.store.select(fromRootReducer.getUserContext);
  }

  get satelliteStyleEnabledText(): string {
    const styleText = this.satelliteStyleEnabled ? 'Map' : 'Satellite';
    return `Toggle ${styleText} View`;
  }

  get mapStyle(): string {
    const style = this.satelliteStyleEnabled ? 'satellite' : 'streets';
    return `mapbox://styles/mapbox/${style}-v9`;
  }

  get center(): any {
    if (!this.map) {
      return [0, 0];
    }
    return this.map.getCenter();
  }

  get pointCompaniesPreview(): string {
    return JSON.parse(this.selectedPoint.properties.Companies).slice(0, 10);
  }

  // Map events
  handleZoomEnd(e: any) {
    this.peerMapInitialZoomComplete$.pipe(take(1)).subscribe(iz => {
      if (!iz) {
        this.store.dispatch(new fromMapActions.InitialZoomComplete());
      }
    });

    this.peerMapAutoZooming$.pipe(take(1)).subscribe(az => {
      if (!!az) {
        this.store.dispatch(new fromMapActions.AutoZoomComplete);
      }
    });
  }

  handleLoadEvent(e: mapboxgl.Map) {
    this.map = e;
    this.store.dispatch(new fromMapActions.MapLoaded());
    /*Quick fix to ensure that events that initialize the map correctly are kicked off*/
    this.handleZoomEnd(null);
    this.handleMoveEndEvent(null);
  }

  handleResizeEvent() {
    this.ignoreNextMoveEnd = true;
  }

  handleMoveEndEvent(e: any) {
    if (!this.ignoreNextMoveEnd) {
      let bounds;
      let zoom;
      // On the initial move end event we want the data to be loaded with the initial map bounds provided to the
      // map without the padding applied. The bounds provided by 'e' in the event have the padding pre applied.
      // This causes data to be loaded in the buffer when loading existing cuts. [BG]
      if (this.initialMoveEnd && this.peerInitialMapBounds && this.peerInitialMapZoomLevel) {
        bounds = new LngLatBounds(
          [this.peerInitialMapBounds[0], this.peerInitialMapBounds[1]],
          [this.peerInitialMapBounds[2], this.peerInitialMapBounds[3]]
        );
        zoom = this.peerInitialMapZoomLevel;
      } else {
        if(e) {
          bounds = e.target.getBounds();
          zoom = e.target.getZoom();
        }
      }
      if(bounds && zoom) {
        this.store.dispatch(new fromMapActions.MoveEnd({
          bounds: bounds,
          zoom: zoom
        }));
        this.initialMoveEnd = false;
      }
    }

    this.ignoreNextMoveEnd = false;
  }

  // Map layer events
  handleLayerHoverEvent(e: any) {
    this.selectedPoint = e.features[0];
  }

  handleLayerClusteredClickEvent(e: any) {
    e.target.flyTo({
      center: e.features[0].geometry.coordinates,
      zoom: e.target.getZoom() + 1
    });
  }

  toggleSatelliteStyle() {
    this.satelliteStyleEnabled = !this.satelliteStyleEnabled;
  }

  ngOnInit(): void {
    this.peerMapSummarySub = this.peerMapSummary$.subscribe(pms => this.peerMapSummary = pms);
    this.peerInitialMapBoundsSub = this.peerInitialMapBounds$.subscribe(ib => this.peerInitialMapBounds = ib);
    this.peerInitialMapZoomLevelSub = this.peerMapInitialZoomLevel$.subscribe(iz => this.peerInitialMapZoomLevel = iz);
    this.userContextSub = this.userContext$.subscribe(uc => {
      this.mbAccessToken = uc.MapboxAccessToken;
    });
  }

  ngOnDestroy() {
    this.peerMapSummarySub.unsubscribe();
    this.peerInitialMapBoundsSub.unsubscribe();
    this.peerInitialMapZoomLevelSub.unsubscribe();
    this.userContextSub.unsubscribe();
  }
}
