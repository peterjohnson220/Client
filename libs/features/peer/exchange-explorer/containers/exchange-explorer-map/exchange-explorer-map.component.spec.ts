import {  NO_ERRORS_SCHEMA } from '@angular/core';
import { TestBed, ComponentFixture } from '@angular/core/testing';

import { StoreModule, Store, combineReducers } from '@ngrx/store';

import * as fromRootState from 'libs/state/state';

import * as fromFeaturePeerMapReducer from '../../reducers';
import { ExchangeExplorerMapComponent } from './exchange-explorer-map.component';

jest.mock('mapbox-gl/dist/mapbox-gl', () => ({
  LngLatBounds: () => ({})
}));

describe('Features - Peer - Map Component', () => {
  let fixture: ComponentFixture<ExchangeExplorerMapComponent>;
  let instance: ExchangeExplorerMapComponent;
  let store: Store<fromRootState.State>;

  // Configure Testing Module for before each test
  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          ...fromRootState.reducers,
          feature_peerMap: combineReducers(fromFeaturePeerMapReducer.reducers)
        })
      ],
      declarations: [
        ExchangeExplorerMapComponent
      ],
      // Shallow Testing
      schemas: [ NO_ERRORS_SCHEMA ]
    });

    store = TestBed.inject(Store);

    fixture = TestBed.createComponent(ExchangeExplorerMapComponent);
    instance = fixture.componentInstance;
  });

  // TODO: Add Tests
  it('does not have tests yet', () => {
    expect(true).toBe(true);
  });

});
