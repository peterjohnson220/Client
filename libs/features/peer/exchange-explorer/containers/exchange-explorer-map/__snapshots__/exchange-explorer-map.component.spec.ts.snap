// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features - Peer - Map Component should load the exchange explorer map with default options 1`] = `
<pf-exchange-explorer-map
  __ngContext__={[Function LRootView]}
  dataLoading$={[Function Store]}
  ignoreNextMoveEnd="false"
  initialMoveEnd={[Function Boolean]}
  peerInitialMapBounds="undefined"
  peerInitialMapBounds$={[Function Store]}
  peerInitialMapBoundsSub={[Function Subscriber]}
  peerInitialMapZoomLevel="undefined"
  peerInitialMapZoomLevelSub={[Function Subscriber]}
  peerMapAutoZooming$={[Function Store]}
  peerMapCentroid$={[Function Store]}
  peerMapCollection$={[Function Store]}
  peerMapHasUntaggedIncumbents$={[Function Store]}
  peerMapInitialZoomComplete$={[Function Store]}
  peerMapInitialZoomLevel$={[Function Store]}
  peerMapLoaded$={[Function Store]}
  peerMapLoadingError$={[Function Store]}
  peerMapMaxZoom$={[Function Store]}
  peerMapShowNoData$={[Function Store]}
  peerMapSummary="undefined"
  peerMapSummary$={[Function Store]}
  peerMapSummarySub={[Function Subscriber]}
  satelliteStyleEnabled="false"
  selectedPoint="null"
  store={[Function Store]}
  userContext$={[Function Store]}
  userContextSub={[Function Subscriber]}
>
  <mgl-map>
    <mgl-control
      mglgeocoder=""
    />
    <mgl-control
      mglnavigation=""
    />
    <mgl-control>
      <div
        class="mapboxgl-ctrl-group"
      >
        <button
          class="btn btn-secondary"
          title="Toggle Satellite View"
        >
          <span>
            <fa-icon
              icon="globe"
            />
          </span>
        </button>
      </div>
    </mgl-control>
    <mgl-geojson-source
      id="incumbents"
    />
    <mgl-layer
      id="multiple-company-latlon-layer"
      source="incumbents"
      type="circle"
    />
    <mgl-layer
      id="single-company-latlon-layer"
      source="incumbents"
      type="circle"
    />
    <mgl-layer
      id="cluster-count-layer"
      source="incumbents"
      type="symbol"
    />
  </mgl-map>
</pf-exchange-explorer-map>
`;

exports[`Features - Peer - Map Component should show the no data message on the map 1`] = `
<pf-exchange-explorer-map
  __ngContext__={[Function LRootView]}
  dataLoading$={[Function Observable]}
  ignoreNextMoveEnd="false"
  initialMoveEnd={[Function Boolean]}
  peerInitialMapBounds="undefined"
  peerInitialMapBounds$={[Function Store]}
  peerInitialMapBoundsSub={[Function Subscriber]}
  peerInitialMapZoomLevel="undefined"
  peerInitialMapZoomLevelSub={[Function Subscriber]}
  peerMapAutoZooming$={[Function Observable]}
  peerMapCentroid$={[Function Store]}
  peerMapCollection$={[Function Store]}
  peerMapHasUntaggedIncumbents$={[Function Store]}
  peerMapInitialZoomComplete$={[Function Store]}
  peerMapInitialZoomLevel$={[Function Store]}
  peerMapLoaded$={[Function Observable]}
  peerMapLoadingError$={[Function Store]}
  peerMapMaxZoom$={[Function Store]}
  peerMapShowNoData$={[Function Observable]}
  peerMapSummary="undefined"
  peerMapSummary$={[Function Store]}
  peerMapSummarySub={[Function Subscriber]}
  satelliteStyleEnabled="false"
  selectedPoint="null"
  store={[Function Store]}
  userContext$={[Function Store]}
  userContextSub={[Function Subscriber]}
>
  <mgl-map
    class="map-no-data"
  >
    <mgl-control
      mglgeocoder=""
    />
    <mgl-control
      mglnavigation=""
    />
    <mgl-control>
      <div
        class="mapboxgl-ctrl-group"
      >
        <button
          class="btn btn-secondary"
          title="Toggle Satellite View"
        >
          <span>
            <fa-icon
              icon="globe"
            />
          </span>
        </button>
      </div>
    </mgl-control>
    <mgl-geojson-source
      id="incumbents"
    />
    <mgl-layer
      id="multiple-company-latlon-layer"
      source="incumbents"
      type="circle"
    />
    <mgl-layer
      id="single-company-latlon-layer"
      source="incumbents"
      type="circle"
    />
    <mgl-layer
      id="cluster-count-layer"
      source="incumbents"
      type="symbol"
    />
    <mgl-popup>
      <div
        class="map-no-data-message"
      >
      </div>
    </mgl-popup>
  </mgl-map>
</pf-exchange-explorer-map>
`;

exports[`Features - Peer - Map Component should show the satellite style if it is enabled 1`] = `
<pf-exchange-explorer-map
  __ngContext__={[Function LRootView]}
  dataLoading$={[Function Store]}
  ignoreNextMoveEnd="false"
  initialMoveEnd={[Function Boolean]}
  peerInitialMapBounds="undefined"
  peerInitialMapBounds$={[Function Store]}
  peerInitialMapBoundsSub={[Function Subscriber]}
  peerInitialMapZoomLevel="undefined"
  peerInitialMapZoomLevelSub={[Function Subscriber]}
  peerMapAutoZooming$={[Function Store]}
  peerMapCentroid$={[Function Store]}
  peerMapCollection$={[Function Store]}
  peerMapHasUntaggedIncumbents$={[Function Store]}
  peerMapInitialZoomComplete$={[Function Store]}
  peerMapInitialZoomLevel$={[Function Store]}
  peerMapLoaded$={[Function Store]}
  peerMapLoadingError$={[Function Store]}
  peerMapMaxZoom$={[Function Store]}
  peerMapShowNoData$={[Function Store]}
  peerMapSummary="undefined"
  peerMapSummary$={[Function Store]}
  peerMapSummarySub={[Function Subscriber]}
  satelliteStyleEnabled={[Function Boolean]}
  selectedPoint="null"
  store={[Function Store]}
  userContext$={[Function Store]}
  userContextSub={[Function Subscriber]}
>
  <mgl-map>
    <mgl-control
      mglgeocoder=""
    />
    <mgl-control
      mglnavigation=""
    />
    <mgl-control>
      <div
        class="mapboxgl-ctrl-group"
      >
        <button
          class="btn btn-secondary"
          title="Toggle Map View"
        >
          <span>
            <fa-icon />
          </span>
        </button>
      </div>
    </mgl-control>
    <mgl-geojson-source
      id="incumbents"
    />
    <mgl-layer
      id="multiple-company-latlon-layer"
      source="incumbents"
      type="circle"
    />
    <mgl-layer
      id="single-company-latlon-layer"
      source="incumbents"
      type="circle"
    />
    <mgl-layer
      id="cluster-count-layer"
      source="incumbents"
      type="symbol"
    />
  </mgl-map>
</pf-exchange-explorer-map>
`;
